File: src/clients/java/src/test/java/com/tigerbeetle/ConformanceTest.java
package com.tigerbeetle;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.xml.parsers.DocumentBuilderFactory;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

@RunWith(Parameterized.class)
public class ConformanceTest {
    private static final byte[] clusterId = new byte[16];

    private static IntegrationTest.Server server;
    private static Client client;

    @BeforeClass
    public static void initialize() throws Exception {
        server = new IntegrationTest.Server("conformance");
        client = new Client(clusterId, new String[] {server.getAddress()});
    }

    @AfterClass
    public static void cleanup() throws Exception {
        client.close();
        server.close();
    }

    @Parameterized.Parameters(name = "{0}")
    public static List<Object[]> cases() throws Exception {
        final var fixture = DocumentBuilderFactory.newInstance().newDocumentBuilder()
                .parse(ConformanceTest.class.getResourceAsStream("/conformance.xml"));

        final var cases = new ArrayList<Object[]>();
        for (final var suite : children(fixture.getDocumentElement())) {
            final var suiteName = child(suite, "suite").getTextContent();
            for (final var testCase : children(child(suite, "cases"))) {
                final var description = child(testCase, "description").getTextContent();
                cases.add(new Object[] {suiteName + ": " + description, testCase});
            }
        }
        return cases;
    }

    private final Element testCase;
    private final Map<Long, byte[]> ids = new HashMap<>();
    private final Map<String, Object> vars = new HashMap<>();

    public ConformanceTest(final String description, final Element testCase) {
        this.testCase = testCase;
    }

    @Test
    public void test() throws Exception {
        final var arrange = childOptional(testCase, "arrange");
        if (arrange != null) {
            for (final var item : children(arrange)) {
                executeArrange(onlyChild(item));
            }
        }

        final var assertions = new ArrayList<Element>();
        for (final var item : children(child(testCase, "assert"))) {
            assertions.add(onlyChild(item));
        }

        if (expectsFail(assertions)) {
            try {
                executeAct();
            } catch (Exception exception) {
                return;
            }
            fail("expected failure");
        } else {
            assertResults(assertions, executeAct());
        }
    }

    private void executeArrange(final Element step) throws Exception {
        switch (step.getTagName()) {
            case "assign":
                vars.put(child(step, "name").getTextContent(), buildValue(child(step, "value")));
                break;
            case "operation":
                executeOperation(step);
                break;
            default:
                fail("unknown arrange step: <" + step.getTagName() + ">");
        }
    }

    private List<Object> executeAct() throws Exception {
        final var results = new ArrayList<Object>();
        for (final var item : children(child(testCase, "act"))) {
            final var step = onlyChild(item);
            if (!step.getTagName().equals("operation")) {
                fail("act step is not an operation: <" + step.getTagName() + ">");
            }
            results.add(executeOperation(step));
        }
        return results;
    }

    private Object executeOperation(final Element operation) throws Exception {
        final var name = child(operation, "name").getTextContent();
        final var inputs = new ArrayList<Object>();
        for (final var item : children(child(operation, "inputs"))) {
            inputs.add(resolveInput(item));
        }

        switch (name) {
            case "create_accounts": {
                final var accounts = new AccountBatch(inputs.size());
                for (final var input : inputs) {
                    addAccount(accounts, elementValue(input, "account"));
                }
                return client.createAccounts(accounts);
            }
            case "lookup_accounts": {
                final var lookupIds = new IdBatch(inputs.size());
                for (final var input : inputs) {
                    lookupIds.add(uint128Value(input));
                }
                return client.lookupAccounts(lookupIds);
            }
            case "create_transfers": {
                final var transfers = new TransferBatch(inputs.size());
                for (final var input : inputs) {
                    addTransfer(transfers, elementValue(input, "transfer"));
                }
                return client.createTransfers(transfers);
            }
            case "lookup_transfers": {
                final var lookupIds = new IdBatch(inputs.size());
                for (final var input : inputs) {
                    lookupIds.add(uint128Value(input));
                }
                return client.lookupTransfers(lookupIds);
            }
            case "get_account_transfers":
                return client.getAccountTransfers(accountFilterValue(inputs));
            case "get_account_balances":
                return client.getAccountBalances(accountFilterValue(inputs));
            case "query_accounts":
                return client.queryAccounts(queryFilterValue(inputs));
            case "query_transfers":
                return client.queryTransfers(queryFilterValue(inputs));
            default:
                fail("unknown operation: " + name);
                return null;
        }
    }

    private Object resolveInput(final Element input) {
        final var from = childOptional(input, "from");
        if (from != null) {
            final var variable = vars.get(from.getTextContent());
            if (variable == null) {
                fail("unknown variable: " + from.getTextContent());
            }
            return variable;
        }

        final var value = childOptional(input, "value");
        if (value != null) {
            return buildValue(value);
        }

        fail("unknown input: <" + input.getTagName() + ">");
        return null;
    }

    private Object buildValue(final Element value) {
        final var element = onlyChild(value);
        switch (element.getTagName()) {
            case "account":
            case "transfer":
                return element;
            case "account_filter":
                return buildAccountFilter(element);
            case "query_filter":
                return buildQueryFilter(element);
            case "id":
                return logicalId(Long.parseLong(element.getTextContent()));
            case "u128":
                return UInt128.asBytes(new BigInteger(element.getTextContent()));
            default:
                fail("unknown value: <" + element.getTagName() + ">");
                return null;
        }
    }

    private void addAccount(final AccountBatch accounts, final Element account) throws Exception {
        accounts.add();
        accounts.setId(uint128Value(buildValue(child(account, "id"))));

        final var userData128 = childOptional(account, "user_data_128");
        if (userData128 != null) {
            accounts.setUserData128(uint128Value(buildValue(userData128)));
        }

        final var ledger = childOptional(account, "ledger");
        if (ledger != null) {
            accounts.setLedger(Integer.parseInt(ledger.getTextContent()));
        }

        final var code = childOptional(account, "code");
        if (code != null) {
            accounts.setCode(Integer.parseInt(code.getTextContent()));
        }

        final var flags = childOptional(account, "flags");
        if (flags != null) {
            accounts.setFlags(flagsValue(flags, AccountFlags.class));
        }
    }

    private AccountFilter buildAccountFilter(final Element element) {
        final var filter = new AccountFilter();
        filter.setAccountId(uint128Value(buildValue(child(element, "account_id"))));

        final var limit = childOptional(element, "limit");
        if (limit != null) {
            filter.setLimit(Integer.parseInt(limit.getTextContent()));
        }

        final var flags = childOptional(element, "flags");
        if (flags != null) {
            try {
                filter.setFlags(flagsValue(flags, AccountFilterFlags.class), true);
            } catch (Exception exception) {
                fail(exception.toString());
            }
        }

        return filter;
    }

    private QueryFilter buildQueryFilter(final Element element) {
        final var filter = new QueryFilter();
        filter.setUserData128(uint128Value(buildValue(child(element, "user_data_128"))));

        final var limit = childOptional(element, "limit");
        if (limit != null) {
            filter.setLimit(Integer.parseInt(limit.getTextContent()));
        }

        final var flags = childOptional(element, "flags");
        if (flags != null) {
            try {
                filter.setFlags(flagsValue(flags, QueryFilterFlags.class), true);
            } catch (Exception exception) {
                fail(exception.toString());
            }
        }

        return filter;
    }

    private static int flagsValue(final Element flags, final Class<?> flagsClass) throws Exception {
        var value = 0;
        for (final var flag : children(flags)) {
            value |= flagsClass.getField(flag.getTextContent().toUpperCase()).getInt(null);
        }
        return value;
    }

    private static AccountFilter accountFilterValue(final List<Object> inputs) {
        if (inputs.size() != 1 || !(inputs.get(0) instanceof AccountFilter)) {
            fail("expected a single account filter: " + inputs);
        }
        return (AccountFilter) inputs.get(0);
    }

    private static QueryFilter queryFilterValue(final List<Object> inputs) {
        if (inputs.size() != 1 || !(inputs.get(0) instanceof QueryFilter)) {
            fail("expected a single query filter: " + inputs);
        }
        return (QueryFilter) inputs.get(0);
    }

    private void addTransfer(final TransferBatch transfers, final Element transfer) {
        transfers.add();
        transfers.setId(uint128Value(buildValue(child(transfer, "id"))));
        transfers.setDebitAccountId(uint128Value(buildValue(child(transfer, "debit_account_id"))));
        transfers
                .setCreditAccountId(uint128Value(buildValue(child(transfer, "credit_account_id"))));

        final var amount = childOptional(transfer, "amount");
        if (amount != null) {
            transfers.setAmount(new BigInteger(amount.getTextContent()));
        }

        final var userData128 = childOptional(transfer, "user_data_128");
        if (userData128 != null) {
            transfers.setUserData128(uint128Value(buildValue(userData128)));
        }

        final var ledger = childOptional(transfer, "ledger");
        if (ledger != null) {
            transfers.setLedger(Integer.parseInt(ledger.getTextContent()));
        }

        final var code = childOptional(transfer, "code");
        if (code != null) {
            transfers.setCode(Integer.parseInt(code.getTextContent()));
        }
    }

    private byte[] logicalId(final long value) {
        return ids.computeIfAbsent(value, unused -> UInt128.id());
    }

    private static Element elementValue(final Object value, final String tag) {
        if (!(value instanceof Element) || !((Element) value).getTagName().equals(tag)) {
            fail("expected <" + tag + ">: " + value);
        }
        return (Element) value;
    }

    private static byte[] uint128Value(final Object value) {
        if (!(value instanceof byte[])) {
            fail("expected a 128-bit value: " + value);
        }
        return (byte[]) value;
    }

    private static boolean expectsFail(final List<Element> assertions) {
        for (final var assertion : assertions) {
            if (assertion.getTagName().equals("assert_fail")) {
                return true;
            }
        }
        return false;
    }

    private void assertResults(final List<Element> assertions, final List<Object> results)
            throws Exception {
        assertEquals("result steps", assertions.size(), results.size());
        for (int i = 0; i < assertions.size(); i++) {
            assertResult(assertions.get(i), results.get(i));
        }
    }

    private void assertResult(final Element assertion, final Object result) throws Exception {
        switch (assertion.getTagName()) {
            case "assert_empty":
                assertEquals(assertion.getTextContent(), resultName(result));
                assertEquals(0, ((Batch) result).getLength());
                break;
            case "assert_equal": {
                assertEquals(child(assertion, "actual").getTextContent(), resultName(result));
                final var expected = children(child(assertion, "expected"));
                if (result instanceof CreateAccountResultBatch) {
                    assertCreateAccountResults(expected, (CreateAccountResultBatch) result);
                } else if (result instanceof CreateTransferResultBatch) {
                    assertCreateTransferResults(expected, (CreateTransferResultBatch) result);
                } else {
                    assertRecords(expected, (Batch) result);
                }
                break;
            }
            default:
                fail("unknown assertion: <" + assertion.getTagName() + ">");
        }
    }

    private static String resultName(final Object result) {
        if (result instanceof CreateAccountResultBatch) {
            return "create_account_results";
        }
        if (result instanceof AccountBatch) {
            return "accounts";
        }
        if (result instanceof CreateTransferResultBatch) {
            return "create_transfer_results";
        }
        if (result instanceof TransferBatch) {
            return "transfers";
        }
        if (result instanceof AccountBalanceBatch) {
            return "account_balances";
        }

        fail("unknown result type: " + result);
        return null;
    }

    private void assertCreateAccountResults(final List<Element> expected,
            final CreateAccountResultBatch actual) {
        assertEquals("create account results", expected.size(), actual.getLength());
        for (final var item : expected) {
            assertTrue(actual.next());
            assertEquals(createAccountStatus(child(item, "status").getTextContent()),
                    actual.getStatus());
        }
    }

    private void assertCreateTransferResults(final List<Element> expected,
            final CreateTransferResultBatch actual) {
        assertEquals("create transfer results", expected.size(), actual.getLength());
        for (final var item : expected) {
            assertTrue(actual.next());
            assertEquals(createTransferStatus(child(item, "status").getTextContent()),
                    actual.getStatus());
        }
    }

    private void assertRecords(final List<Element> expected, final Batch actual) throws Exception {
        assertEquals("records", expected.size(), actual.getLength());
        for (final var item : expected) {
            assertTrue(actual.next());
            for (final var field : children(item)) {
                final var actualValue =
                        actual.getClass().getMethod(getterName(field.getTagName())).invoke(actual);
                if (children(field).isEmpty()) {
                    assertEquals(field.getTagName(), Long.parseLong(field.getTextContent()),
                            ((Number) actualValue).longValue());
                } else {
                    assertArrayEquals(uint128Value(buildValue(field)), (byte[]) actualValue);
                }
            }
        }
    }

    // The fixture's snake_case field maps to a getter, e.g. user_data_128 to getUserData128().
    private static String getterName(final String field) {
        final var name = new StringBuilder("get");
        for (final var part : field.split("_")) {
            name.append(Character.toUpperCase(part.charAt(0))).append(part.substring(1));
        }
        return name.toString();
    }

    private static CreateAccountStatus createAccountStatus(final String status) {
        switch (status) {
            case "created":
                return CreateAccountStatus.Created;
            case "exists":
                return CreateAccountStatus.Exists;
            default:
                fail("unknown create account status: " + status);
                return null;
        }
    }

    private static CreateTransferStatus createTransferStatus(final String status) {
        switch (status) {
            case "created":
                return CreateTransferStatus.Created;
            case "exists":
                return CreateTransferStatus.Exists;
            default:
                fail("unknown create transfer status: " + status);
                return null;
        }
    }

    private static List<Element> children(final Element parent) {
        final var elements = new ArrayList<Element>();
        final var nodes = parent.getChildNodes();
        for (int i = 0; i < nodes.getLength(); i++) {
            if (nodes.item(i).getNodeType() == Node.ELEMENT_NODE) {
                elements.add((Element) nodes.item(i));
            }
        }
        return elements;
    }

    private static Element child(final Element parent, final String tag) {
        final var element = childOptional(parent, tag);
        if (element == null) {
            fail("missing <" + tag + "> in <" + parent.getTagName() + ">");
        }
        return element;
    }

    private static Element childOptional(final Element parent, final String tag) {
        for (final var element : children(parent)) {
            if (element.getTagName().equals(tag)) {
                return element;
            }
        }
        return null;
    }

    // Steps, values, and assertions are encoded as an element with exactly one child.
    private static Element onlyChild(final Element parent) {
        final var elements = children(parent);
        if (elements.size() != 1) {
            fail("expected exactly one child in <" + parent.getTagName() + ">");
        }
        return elements.get(0);
    }
}

File: src/clients/go/conformance_test.go
package tigerbeetle_go

import (
	"bytes"
	"encoding/json"
	"errors"
	"math/big"
	"os"
	"reflect"
	"strings"
	"testing"
)

var errInvalidConformanceStep = errors.New("invalid conformance step")

type conformanceFixture struct {
	Suites []conformanceSuite `json:"suites"`
}

type conformanceSuite struct {
	Suite string            `json:"suite"`
	Cases []conformanceCase `json:"cases"`
}

type conformanceCase struct {
	Description string            `json:"description"`
	Arrange     []conformanceStep `json:"arrange"`
	Act         []conformanceStep `json:"act"`
	Assert      []json.RawMessage `json:"assert"`
}

type conformanceStep struct {
	Type      string
	Assign    conformanceAssign
	Operation conformanceOperation
}

func (step *conformanceStep) UnmarshalJSON(data []byte) error {
	var tagged map[string]json.RawMessage
	if err := json.Unmarshal(data, &tagged); err != nil {
		return err
	}
	if len(tagged) != 1 {
		return errInvalidConformanceStep
	}

	for stepType, value := range tagged {
		step.Type = stepType
		switch stepType {
		case "assign":
			return json.Unmarshal(value, &step.Assign)
		case "operation":
			return json.Unmarshal(value, &step.Operation)
		default:
			return errInvalidConformanceStep
		}
	}

	return errInvalidConformanceStep
}

func decodeTaggedObject(t testing.TB, data json.RawMessage) (string, json.RawMessage) {
	var tagged map[string]json.RawMessage
	decodeJSON(t, data, &tagged)
	if len(tagged) != 1 {
		t.Fatalf("expected exactly one tag: %s", data)
	}

	for tag, value := range tagged {
		return tag, value
	}

	t.Fatalf("expected exactly one tag: %s", data)
	return "", nil
}

func decodeJSON(t testing.TB, data json.RawMessage, value interface{}) {
	if err := json.Unmarshal(data, value); err != nil {
		t.Fatal(err)
	}
}

type conformanceAssign struct {
	Name  string          `json:"name"`
	Value json.RawMessage `json:"value"`
}

type conformanceOperation struct {
	Name   string             `json:"name"`
	Inputs []conformanceInput `json:"inputs"`
}

type conformanceInput struct {
	From  string          `json:"from"`
	Value json.RawMessage `json:"value"`
}

type conformanceContext struct {
	ids  map[uint64]Uint128
	vars map[string]interface{}
}

func TestConformance(t *testing.T) {
	fixtureBytes, err := os.ReadFile("conformance/conformance.json")
	if err != nil {
		t.Fatal(err)
	}

	var fixture conformanceFixture
	if err := json.Unmarshal(fixtureBytes, &fixture); err != nil {
		t.Fatal(err)
	}

	WithClient(t, func(client Client) {
		for _, suite := range fixture.Suites {
			for _, testCase := range suite.Cases {
				description := suite.Suite + ": " + testCase.Description
				t.Run(description, func(t *testing.T) {
					context := conformanceContext{
						ids:  map[uint64]Uint128{},
						vars: map[string]interface{}{},
					}

					for _, step := range testCase.Arrange {
						executeArrange(t, client, step, context)
					}

					if expectsFail(t, testCase.Assert) {
						assertFails(t, description, func() error {
							_, err := executeAct(t, client, testCase, context)
							return err
						})
					} else {
						results, err := executeAct(t, client, testCase, context)
						if err != nil {
							t.Fatalf("%s: %s", description, err)
						}
						assertResults(t, testCase.Assert, results, description, context)
					}
				})
			}
		}
	})
}

func executeAct(
	t testing.TB,
	client Client,
	testCase conformanceCase,
	context conformanceContext,
) ([]interface{}, error) {
	results := make([]interface{}, len(testCase.Act))
	for i, step := range testCase.Act {
		if step.Type != "operation" {
			t.Fatalf("act step is not an operation: %#v", step)
		}
		result, err := executeOperation(t, client, step.Operation, context)
		if err != nil {
			return nil, err
		}
		results[i] = result
	}
	return results, nil
}

func expectsFail(t testing.TB, assertions []json.RawMessage) bool {
	for _, assertion := range assertions {
		tag, _ := decodeTaggedObject(t, assertion)
		if tag == "assert_fail" {
			return true
		}
	}
	return false
}

// A failure is either an error returned by the client or a panic.
func assertFails(t testing.TB, description string, action func() error) {
	failed := false
	func() {
		defer func() {
			if recover() != nil {
				failed = true
			}
		}()
		failed = action() != nil
	}()

	if !failed {
		t.Fatalf("%s: expected failure", description)
	}
}

func executeArrange(t testing.TB, client Client, step conformanceStep, context conformanceContext) {
	switch step.Type {
	case "assign":
		assign(t, step.Assign, context)
	case "operation":
		if _, err := executeOperation(t, client, step.Operation, context); err != nil {
			t.Fatalf("arrange operation failed: %s", err)
		}
	default:
		t.Fatalf("unknown arrange step: %#v", step)
	}
}

func assign(t testing.TB, assign conformanceAssign, context conformanceContext) {
	context.vars[assign.Name] = buildValue(t, assign.Value, context)
}

func executeOperation(
	t testing.TB,
	client Client,
	operation conformanceOperation,
	context conformanceContext,
) (interface{}, error) {
	inputs := make([]interface{}, len(operation.Inputs))
	for i, input := range operation.Inputs {
		inputs[i] = resolveInput(t, input, context)
	}

	switch operation.Name {
	case "create_accounts":
		accounts := make([]Account, len(inputs))
		for i, input := range inputs {
			account, ok := input.(Account)
			if !ok {
				t.Fatalf("create_accounts input is not an Account: %#v", input)
			}
			accounts[i] = account
		}

		return client.CreateAccounts(accounts)
	case "lookup_accounts":
		ids := make([]Uint128, len(inputs))
		for i, input := range inputs {
			ids[i] = uint128Value(t, input)
		}

		return client.LookupAccounts(ids)
	case "create_transfers":
		transfers := make([]Transfer, len(inputs))
		for i, input := range inputs {
			transfer, ok := input.(Transfer)
			if !ok {
				t.Fatalf("create_transfers input is not a Transfer: %#v", input)
			}
			transfers[i] = transfer
		}

		return client.CreateTransfers(transfers)
	case "lookup_transfers":
		ids := make([]Uint128, len(inputs))
		for i, input := range inputs {
			ids[i] = uint128Value(t, input)
		}

		return client.LookupTransfers(ids)
	case "get_account_transfers":
		return client.GetAccountTransfers(accountFilterValue(t, inputs))
	case "get_account_balances":
		return client.GetAccountBalances(accountFilterValue(t, inputs))
	default:
		t.Fatalf("unknown operation: %s", operation.Name)
		return nil, nil
	}
}

func resolveInput(t testing.TB, input conformanceInput, context conformanceContext) interface{} {
	if input.From != "" {
		value, ok := context.vars[input.From]
		if !ok {
			t.Fatalf("unknown variable: %s", input.From)
		}
		return value
	}
	if input.Value != nil {
		return buildValue(t, input.Value, context)
	}

	t.Fatalf("unknown input: %#v", input)
	return nil
}

func buildValue(t testing.TB, value json.RawMessage, context conformanceContext) interface{} {
	tag, data := decodeTaggedObject(t, value)

	switch tag {
	case "account":
		return buildAccount(t, data, context)
	case "transfer":
		return buildTransfer(t, data, context)
	case "account_filter":
		return buildAccountFilter(t, data, context)
	case "id":
		var id uint64
		decodeJSON(t, data, &id)
		return logicalID(id, context)
	case "u128":
		var value string
		decodeJSON(t, data, &value)
		return u128(t, value)
	default:
		t.Fatalf("unknown value: %s", tag)
		return nil
	}
}

func u128(t testing.TB, value string) Uint128 {
	integer, ok := new(big.Int).SetString(value, 10)
	if !ok {
		t.Fatalf("invalid u128: %s", value)
	}
	return BigIntToUint128(integer)
}

func buildAccount(t testing.TB, data json.RawMessage, context conformanceContext) Account {
	var account struct {
		ID     json.RawMessage `json:"id"`
		Ledger uint32          `json:"ledger"`
		Code   uint16          `json:"code"`
		Flags  []string        `json:"flags"`
	}
	decodeJSON(t, data, &account)

	var flags AccountFlags
	buildFlags(t, account.Flags, &flags)

	return Account{
		ID:     uint128Value(t, buildValue(t, account.ID, context)),
		Ledger: account.Ledger,
		Code:   account.Code,
		Flags:  flags.ToUint16(),
	}
}

func buildAccountFilter(
	t testing.TB,
	data json.RawMessage,
	context conformanceContext,
) AccountFilter {
	var filter struct {
		AccountID json.RawMessage `json:"account_id"`
		Limit     uint32          `json:"limit"`
		Flags     []string        `json:"flags"`
	}
	decodeJSON(t, data, &filter)

	var flags AccountFilterFlags
	buildFlags(t, filter.Flags, &flags)

	return AccountFilter{
		AccountID: uint128Value(t, buildValue(t, filter.AccountID, context)),
		Limit:     filter.Limit,
		Flags:     flags.ToUint32(),
	}
}

// Sets the named boolean fields on a client flags struct, e.g. AccountFlags.
func buildFlags(t testing.TB, names []string, flags interface{}) {
	value := reflect.ValueOf(flags).Elem()
	for _, name := range names {
		field := recordField(value, name)
		if !field.IsValid() {
			t.Fatalf("unknown flag: %s", name)
		}
		field.SetBool(true)
	}
}

func buildTransfer(t testing.TB, data json.RawMessage, context conformanceContext) Transfer {
	var transfer struct {
		ID              json.RawMessage `json:"id"`
		DebitAccountID  json.RawMessage `json:"debit_account_id"`
		CreditAccountID json.RawMessage `json:"credit_account_id"`
		Amount          uint64          `json:"amount"`
		Ledger          uint32          `json:"ledger"`
		Code            uint16          `json:"code"`
	}
	decodeJSON(t, data, &transfer)

	return Transfer{
		ID:              uint128Value(t, buildValue(t, transfer.ID, context)),
		DebitAccountID:  uint128Value(t, buildValue(t, transfer.DebitAccountID, context)),
		CreditAccountID: uint128Value(t, buildValue(t, transfer.CreditAccountID, context)),
		Amount:          ToUint128(transfer.Amount),
		Ledger:          transfer.Ledger,
		Code:            transfer.Code,
	}
}

func uint128Value(t testing.TB, value interface{}) Uint128 {
	u128, ok := value.(Uint128)
	if !ok {
		t.Fatalf("expected Uint128: %#v", value)
	}
	return u128
}

func accountFilterValue(t testing.TB, inputs []interface{}) AccountFilter {
	if len(inputs) != 1 {
		t.Fatalf("expected a single account filter, got %d inputs", len(inputs))
	}
	filter, ok := inputs[0].(AccountFilter)
	if !ok {
		t.Fatalf("expected AccountFilter: %#v", inputs[0])
	}
	return filter
}

func logicalID(value uint64, context conformanceContext) Uint128 {
	id, ok := context.ids[value]
	if !ok {
		id = ID()
		context.ids[value] = id
	}
	return id
}

func assertResults(
	t testing.TB,
	expected []json.RawMessage,
	results []interface{},
	description string,
	context conformanceContext,
) {
	if len(expected) != len(results) {
		t.Fatalf("%s: expected %d result steps, got %d", description, len(expected), len(results))
	}

	for i := range expected {
		assertResult(t, expected[i], results[i], description, context)
	}
}

func assertResult(
	t testing.TB,
	assertion json.RawMessage,
	result interface{},
	description string,
	context conformanceContext,
) {
	assertionType, assertionValue := decodeTaggedObject(t, assertion)

	switch assertionType {
	case "assert_empty":
		var expectedName string
		decodeJSON(t, assertionValue, &expectedName)
		assertResultEmpty(t, expectedName, result, description)
	case "assert_equal":
		var equal struct {
			Actual   string          `json:"actual"`
			Expected json.RawMessage `json:"expected"`
		}
		decodeJSON(t, assertionValue, &equal)
		assertResultEqual(t, equal.Actual, equal.Expected, result, description, context)
	default:
		t.Fatalf("%s: unknown assertion type: %s", description, assertionType)
	}
}

func resultName(t testing.TB, result interface{}) (string, int) {
	switch result := result.(type) {
	case []CreateAccountResult:
		return "create_account_results", len(result)
	case []Account:
		return "accounts", len(result)
	case []CreateTransferResult:
		return "create_transfer_results", len(result)
	case []Transfer:
		return "transfers", len(result)
	case []AccountBalance:
		return "account_balances", len(result)
	default:
		t.Fatalf("unknown result type: %#v", result)
		return "", 0
	}
}

func assertResultEmpty(
	t testing.TB,
	expectedName string,
	result interface{},
	description string,
) {
	name, length := resultName(t, result)
	if name != expectedName {
		t.Fatalf("%s: expected %s, got %s", description, expectedName, name)
	}
	if length != 0 {
		t.Fatalf("%s: expected empty %s, got %d", description, name, length)
	}
}

func assertResultEqual(
	t testing.TB,
	expectedName string,
	expectedValue json.RawMessage,
	result interface{},
	description string,
	context conformanceContext,
) {
	name, _ := resultName(t, result)
	if name != expectedName {
		t.Fatalf("%s: expected %s, got %s", description, expectedName, name)
	}

	switch result := result.(type) {
	case []CreateAccountResult:
		assertCreateAccountResults(t, expectedValue, result, description)
	case []CreateTransferResult:
		assertCreateTransferResults(t, expectedValue, result, description)
	default:
		assertRecords(t, expectedValue, result, description, context)
	}
}

func assertCreateAccountResults(
	t testing.TB,
	expectedJSON json.RawMessage,
	actual []CreateAccountResult,
	description string,
) {
	var expected []struct {
		Status string `json:"status"`
	}
	decodeJSON(t, expectedJSON, &expected)

	if len(expected) != len(actual) {
		t.Fatalf(
			"%s: expected %d create account results, got %d",
			description,
			len(expected),
			len(actual),
		)
	}

	for i := range expected {
		status := createAccountStatus(t, expected[i].Status)
		if status != actual[i].Status {
			t.Fatalf(
				"%s: expected create account result %d status %s, got %s",
				description,
				i,
				status,
				actual[i].Status,
			)
		}
	}
}

func assertCreateTransferResults(
	t testing.TB,
	expectedJSON json.RawMessage,
	actual []CreateTransferResult,
	description string,
) {
	var expected []struct {
		Status string `json:"status"`
	}
	decodeJSON(t, expectedJSON, &expected)

	if len(expected) != len(actual) {
		t.Fatalf(
			"%s: expected %d create transfer results, got %d",
			description,
			len(expected),
			len(actual),
		)
	}

	for i := range expected {
		status := createTransferStatus(t, expected[i].Status)
		if status != actual[i].Status {
			t.Fatalf(
				"%s: expected create transfer result %d status %s, got %s",
				description,
				i,
				status,
				actual[i].Status,
			)
		}
	}
}

func assertRecords(
	t testing.TB,
	expectedJSON json.RawMessage,
	actual interface{},
	description string,
	context conformanceContext,
) {
	var expected []map[string]json.RawMessage
	decodeJSON(t, expectedJSON, &expected)

	records := reflect.ValueOf(actual)
	if len(expected) != records.Len() {
		t.Fatalf("%s: expected %d records, got %d", description, len(expected), records.Len())
	}

	for i := range expected {
		assertRecord(t, expected[i], records.Index(i), description, i, context)
	}
}

func assertRecord(
	t testing.TB,
	expected map[string]json.RawMessage,
	actual reflect.Value,
	description string,
	index int,
	context conformanceContext,
) {
	for field, value := range expected {
		actualField := recordField(actual, field)
		if !actualField.IsValid() {
			t.Fatalf("%s: unknown field: %s", description, field)
		}

		var expectedValue interface{}
		if bytes.HasPrefix(bytes.TrimSpace(value), []byte("{")) {
			expectedValue = buildValue(t, value, context)
		} else if actualField.Type() == reflect.TypeOf(Uint128{}) {
			var number uint64
			decodeJSON(t, value, &number)
			expectedValue = ToUint128(number)
		} else {
			pointer := reflect.New(actualField.Type())
			decodeJSON(t, value, pointer.Interface())
			expectedValue = pointer.Elem().Interface()
		}

		if !reflect.DeepEqual(expectedValue, actualField.Interface()) {
			t.Fatalf(
				"%s: expected record %d %s %v, got %v",
				description,
				index,
				field,
				expectedValue,
				actualField.Interface(),
			)
		}
	}
}

// Matches a fixture field name to a struct field, ignoring case and underscores,
// e.g. user_data_128 to UserData128.
func recordField(record reflect.Value, name string) reflect.Value {
	normalized := strings.ReplaceAll(name, "_", "")
	for i := 0; i < record.NumField(); i++ {
		if strings.EqualFold(record.Type().Field(i).Name, normalized) {
			return record.Field(i)
		}
	}
	return reflect.Value{}
}

func createAccountStatus(t testing.TB, status string) CreateAccountStatus {
	switch status {
	case "created":
		return AccountCreated
	case "exists":
		return AccountExists
	default:
		t.Fatalf("unknown create account status: %s", status)
		return 0
	}
}

func createTransferStatus(t testing.TB, status string) CreateTransferStatus {
	switch status {
	case "created":
		return TransferCreated
	case "exists":
		return TransferExists
	default:
		t.Fatalf("unknown create transfer status: %s", status)
		return 0
	}
}
